src/github.com/matrix-org/dendrite/appservice/storage/appservice_events_table.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2018 New Vector Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package storage
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/matrix-org/gomatrixserverlib"
+	log "github.com/sirupsen/logrus"
+)
+
+const appserviceEventsSchema = `
+-- Stores events to be sent to application services
+CREATE TABLE IF NOT EXISTS appservice_events (
+	-- An auto-incrementing id unique to each event in the table
+	id BIGSERIAL NOT NULL PRIMARY KEY,
+	-- The ID of the application service the event will be sent to
+	as_id TEXT NOT NULL,
+	-- The ID of the event
+	event_id TEXT NOT NULL,
+	-- Unix seconds that the event was sent at from the originating server
+	origin_server_ts BIGINT NOT NULL,
+	-- The ID of the room that the event was sent in
+	room_id TEXT NOT NULL,
+	-- The type of the event (e.g. m.text)
+	type TEXT NOT NULL,
+	-- The ID of the user that sent the event
+	sender TEXT NOT NULL,
+	-- The JSON representation of the event's content. Text to avoid db JSON parsing
+	event_content TEXT,
+	-- The ID of the transaction that this event is a part of
+	txn_id BIGINT NOT NULL
+);
+
+CREATE INDEX IF NOT EXISTS appservice_events_as_id ON appservice_events(as_id);
+`
+
+const selectEventsByApplicationServiceIDSQL = "" +
+	"SELECT id, event_id, origin_server_ts, room_id, type, sender, event_content, txn_id " +
+	"FROM appservice_events WHERE as_id = $1 ORDER BY txn_id DESC LIMIT $2"
```

Comment:
Is it reasonable to change this to `ORDER BY txn_id DESC, id ASC`?

Reply:
Yes, that is supported and will prevent events from going out of order :)